File: Sources/ExcalidrawCollab/Reconcile.swift
import ExcalidrawModel

/// Element reconciliation — the deterministic, symmetric last-writer-wins rule
/// shared bit-for-bit with the TypeScript `@xs/protocol` `reconcile`: higher
/// `version` wins; ties break on the lower `versionNonce`. Both clients converge
/// on the same element with no central authority or CRDT.
public enum Reconcile {
    /// Whether `remote` should replace the same-id `local` element.
    public static func preferRemote(local: ExcalidrawElement, remote: ExcalidrawElement) -> Bool {
        if remote.base.version != local.base.version {
            return remote.base.version > local.base.version
        }
        return remote.base.versionNonce < local.base.versionNonce
    }

    /// The winning element of a same-id pair (the local one on an exact tie).
    public static func reconcile(local: ExcalidrawElement, remote: ExcalidrawElement) -> ExcalidrawElement {
        preferRemote(local: local, remote: remote) ? remote : local
    }

    /// Merge a remote batch into `local` by id, preserving local order and
    /// appending remote-only elements.
    public static func reconcileElements(
        local: [ExcalidrawElement], remote: [ExcalidrawElement]
    ) -> [ExcalidrawElement] {
        var remoteById: [String: ExcalidrawElement] = [:]
        for element in remote {
            remoteById[element.id] = element
        }

        var seen = Set<String>()
        var merged: [ExcalidrawElement] = []
        for localElement in local {
            if let remoteElement = remoteById[localElement.id] {
                merged.append(reconcile(local: localElement, remote: remoteElement))
            } else {
                merged.append(localElement)
            }
            seen.insert(localElement.id)
        }
        for remoteElement in remote where !seen.contains(remoteElement.id) {
            merged.append(remoteElement)
            seen.insert(remoteElement.id)
        }
        return merged
    }

    /// Only the elements that changed (new id, or the remote version won).
    public static func changedByReconcile(
        local: [ExcalidrawElement], remote: [ExcalidrawElement]
    ) -> [ExcalidrawElement] {
        var localById: [String: ExcalidrawElement] = [:]
        for element in local {
            localById[element.id] = element
        }
        return remote.filter { remoteElement in
            guard let localElement = localById[remoteElement.id] else { return true }
            return preferRemote(local: localElement, remote: remoteElement)
        }
    }
}

File: android/core-model/src/main/kotlin/com/excalidraw/model/Reconcile.kt
package com.excalidraw.model

import kotlinx.serialization.json.JsonObject

/**
 * Element reconciliation — the deterministic, symmetric last-writer-wins rule
 * shared bit-for-bit with the Swift (`ExcalidrawCollab.Reconcile`) and
 * TypeScript (`excalidraw-svelte/protocol/reconcile`) clients: higher `version`
 * wins; ties break on the lower `versionNonce`. Both peers pick the same winner
 * with no central authority or CRDT, so a mixed iOS/web/Android room converges.
 */
object Reconcile {

    /** Whether [remote] should replace the same-id [local] element. */
    fun preferRemote(local: JsonObject, remote: JsonObject): Boolean {
        val lv = ElementView(local).version
        val rv = ElementView(remote).version
        if (rv != lv) return rv > lv
        return ElementView(remote).versionNonce < ElementView(local).versionNonce
    }

    fun reconcile(local: JsonObject, remote: JsonObject): JsonObject =
        if (preferRemote(local, remote)) remote else local

    private fun idOf(obj: JsonObject): String = ElementView(obj).id

    /**
     * Merge [remote] into [local] by id, preserving local order (winners
     * substituted in place) and appending remote-only elements in order.
     */
    fun reconcileElements(local: List<JsonObject>, remote: List<JsonObject>): List<JsonObject> {
        val remoteById = remote.associateBy(::idOf)
        val seen = HashSet<String>()
        val merged = ArrayList<JsonObject>(local.size + remote.size)
        for (localEl in local) {
            val id = idOf(localEl)
            val remoteEl = remoteById[id]
            merged.add(if (remoteEl == null) localEl else reconcile(localEl, remoteEl))
            seen.add(id)
        }
        for (remoteEl in remote) {
            if (seen.add(idOf(remoteEl))) merged.add(remoteEl)
        }
        return merged
    }

    /** Only the elements that actually changed (new id, or the remote version won). */
    fun changedByReconcile(local: List<JsonObject>, remote: List<JsonObject>): List<JsonObject> {
        val localById = local.associateBy(::idOf)
        return remote.filter { remoteEl ->
            val localEl = localById[idOf(remoteEl)]
            localEl == null || preferRemote(localEl, remoteEl)
        }
    }
}
